pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,242 @@
+//
+// Flush should be called after the last key is written to ensure that the SST
+// is written to the destination.
+func (s *SSTSinkKeyWriter) WriteKey(ctx context.Context, key storage.MVCCKey, value []byte) error {
+	if len(s.flushedFiles) == 0 {
+		return errors.AssertionFailedf(
+			"no BackupManifest_File to write key to, call Reset before WriteKey",
+		)
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
+	if !lastFile.Span.ContainsKey(key.Key) {
+		return errors.AssertionFailedf(
+			"key %s not in span %s, call Reset on the new span", key.Key, lastFile.Span,
+		)
+	}
+	if s.prevKey != nil && s.prevKey.Compare(key.Key) >= 0 {
+		return errors.AssertionFailedf("key %s must be greater than previous key %s", key.Key, s.prevKey)
+	}
+	// At this point, because of the Reset invariant, we can make the following
+	// assumptions about the new key:
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	lastFile, err := s.maybeDoSizeFlush(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span. It will either open a new
+// BackupManifest_File, or if the new span contiguously extends the last span,
+// it will extend it if it remains under the fileSpanByteLimit. The caller is
+// responsible for ensuring that spans that are reset on do not end mid-row.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	log.VEventf(ctx, 2, "resetting sink to span %s", newSpan)
+	if err := s.maybeFlushNonExtendableSpans(newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	// At this point, we can assume the new span is either contiguous with the
+	// last span or comes after the last span (if it exists).
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		if isContiguousSpan(lastFile.Span, newSpan) &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
```

Comment:
I should have been clearer earlier. Reset() should never get called if the latest key in the last file is mid row. If it is, something is deeply wrong, as it implies we could flush mid row. We should only extend the existing span if we're contiguous and the last file is small enough.

Reply:
Are we leaving the mid-row check as a responsibility of the caller then? Seems a little incongruous with the fact that `Flush` returns an error if you attempt to flush mid-row.

Reply:
See comment above. happy to discuss over a call once you've digested my suggestion.